NEP 40: Informational NEP about current DTypes

seberg opened the pull request:
[Rendered version as of 2020-03-11]( https://github.com/numpy/numpy/blob/1248cf7a8765b7b53d883f9e7061173817533aac/doc/neps/nep-0040-legacy-datatype-impl.rst)

This NEP is meant to be mostly informational. The idea is to start a serious of NEPs:

1. NEP 40: Informational
2. NEP 41: Overview, and limited design decisions (and limited decision to start on an implementation)
3. NEP 42: Technical details of new DTypes (and casting)
4. NEP 43: Technical details of revamped UFuncs.

Where the second two could maybe be split up further or simply expanded with time.

seberg committed:
NEP: Create informational NEP 40 about current DTypes

stefanv reviewed (commented):
Well, I learned a lot—thanks!  This was a first pass read, and it would probably be worth (for all these NEPs) looking at the outlines to see if they make sense too.  Will ping you about that tomorrow.

stefanv reviewed (commented):
Well, I learned a lot—thanks!  This was a first pass read, and it would probably be worth (for all these NEPs) looking at the outlines to see if they make sense too.  Will ping you about that tomorrow.

mattip commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,481 @@
+--------
+
+In preparation of NumPy enhancement proposals 41, 42, and 43, this
+NEP details the current status of NumPy datatypes as of NumPy 1.19.
```
19 -> 18 since I hope we get this all done for 1.19

mattip commented:
I think it would be helpful to use sub-headings to distinguish between "current model", "problems with the current model", and "ideas toward solving the problems".

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,481 @@
+``np.load`` (and others) currently translate all extension dtypes to void dtypes.
+This means they cannot be stored using the ``npy`` format.
+If we wish to add support for ``npy`` format in user datatypes, a new protocol
+will have to be designed to achieve this.
```
Do you mean the byte contents of the array?

We could largely achieve that today if we used `pickle` to serialize the dtype instead of a custom text format. I suspect that using `pickle.dump` instead of `np.save` already does the right thing.

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,481 @@
+Open Issues
+-----------
+
+``np.load`` (and others) currently translate all extension dtypes to void dtypes.
```
Isn't it strictly the saving that does the translation?

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,481 @@
+Even though NumPy itself may never have true support for masked values,
+the datatype system could grow it.
+This means support for datatypes where the actual mask information is not
+stored as part of the datatype, but as part of a bit array in the array object.
```
This seems to be a special case combining:
* having field-clustered storage for structured dtypes, rather than element-clustered (struct of arrays vs array of structs)
* support for bit-strides

Both of these would require fairly fundamental changes to the semantics of `arr.strides` and `arr.view`, which sounds like a can of worms best left unopened.

seberg replied:
Ah, to clarify that. I have no intention to even think about opening the can of worms for numpy. I noted it, because pandas has/will have such types, and if it is much more straight forward, we could think about helping out there.

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,481 @@
+
+.. _safe_casting: Safe casting denotes the concept that the value held by one dtype can be represented by another one without loss/change of information. Within current NumPy there are two slightly different usages. First, casting to string is considered safe, although it is not safe from a type perspective (it is safe in the sense that it cannot fail); this behaviour should be considered legacy. ...
+
+.. _flexible_dtype: A flexible dtype is a dtype for which conversion is not always safely possible. This is for example the case for current string dtypes, which can have different lengths. It is also true for datetime64 due to its attached unit. A non-flexible dtype should typically have a canonical representation (i.e. a float64 may be in non-native byteorder, but the default is native byte ord ...
```
This definition is not consistent with `PyArray_ISFLEXIBLE`, which includes only void, str, and bytes.

seberg commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,481 @@
+Flexible here means the values representable by the datatype depend on the
+specific instance.
+As an example, ``datetime64[ns]`` can represent partial seconds which
+``datetime64[s]`` cannot and ``"S8"`` can represent longer strings than ``"S4"``.
```
Yeah, forgot about `np.flexible` (and ignored the C-API a bit, although it is mixed, e.g. `AdaptFlexible` of course also adepts the "generic" datetime). I guess I need a new word, hmmm.

seberg replied:
Anyway, its incorrect here, you are right. Since it was supposed to describe the current state more. I think "variable lengths" is not a very important distinction for most code.

seberg replied:
I think I will try to go with "parametric" for meaning "there is no canonical instance and different dtype instances may not be able to represent the exact same values". (used here: https://github.com/numpy/numpy/issues/12518#issuecomment-446059956) At least for now, of course "byte order" is also a parameter so to say, but...

seberg commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,481 @@
+1. Object datatypes require reference counting, which adds complexity to
+   which we may not want to (initially) hide from general user defined datatypes.
+2. In many use-cases it seems likely that a generic type specific solution
+   can be sufficient to improve type safety in user-code (e.g. by specifying
```
Removed it, it is mainly a repetition anyway.

seberg committed:
First pass of revising, mainly Stéfans and Erics comments

Next pass will be to reorganize everything a bit as Matti suggested it
there is a very little start here already

seberg committed:
Reorganize sections

Hopefully nothing broke much, but I think so, also removes a bit
of duplication which I had missed before :/

rgommers commented:
It may be worth asking for some help here. E.g. one of the Pandas devs (@TomAugspurger or @jorisvandenbossche any interest, or can you suggest someone?) may be willing to write up one or two examples of where this helps Pandas; @mhvk or @ngoldbaum or @hgrecco may be willing to adapt a units example of astropy.units/unyt/pint where they can see this having value.

mattip committed:
rewording

seberg committed:
Merge pull request #25 from mattip/nep40-2

rewording

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+variable length (string, bytes,
+and void). This distinction is similarly exposed by the C-Macros
+``PyDataType_ISFLEXIBLE`` and ``PyTypeNum_ISFLEXIBLE``.
+For instance, ``"S8"`` can represent longer strings than ``"S4"``.
```
This sentence feels a little out of place here, but probably is sensible to work in somehow.

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+For instance, ``"S8"`` can represent longer strings than ``"S4"``.
+
+The basic numerical datatypes are not flexible (do not inherit from
+``npflexible``). Float64, Float32, etc. do have a byte order, but the described
```
```suggestion
``np.flexible``). ``float64``, ``float32``, etc. do have a byte order, but the described
```

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+The concept of flexibility can be generalized to parametric datatypes.
+For example the private ``AdaptFlexibleDType`` function also accepts the
+naive datetime dtype as input to find the correct time unit.
+The datetime dtype is thus parametric not in the storage, but instead in
```
Perhaps:
```suggestion
The datetime dtype is thus parametric not in the size of it's storage, but instead in
```

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+
+1. Casting is not always safe (``np.can_cast("S8", "S4")``)
+2. Array coercion should be able to discover the exact dtype, such as for
+   ``np.array(["str1", 123.], dtype="S")`` where NumPy discovers the
```
Perhaps `12.34` would make this example clearer

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+2. Array coercion should be able to discover the exact dtype, such as for
+   ``np.array(["str1", 123.], dtype="S")`` where NumPy discovers the
+   resulting dtype as ``"S5"``.
+   (Without ``dtype="S"`` such behaviour is currently ill defined [gh-15327]_.)
```
Do you mean "if the dtype argument is ommitted"? The behavior is not ill-defined if `np.object_` is passed, for instance.

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+will lead to the same results, and thus ignores the existing datatype.
+The same logic also applies to floating point scalars, which are allowed to
+lose precision.
+The behavior is not used when both inputs are scalars.
```
Can you give an example here?

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+do not necessarily work, but can work for a :class:`decimal.Decimal`.
+To improve this situation it seems desirable to make it easy to create
+object dtypes that represent a specific python datatype and stores its object
+inside the array in the form of pointer to python `PyOjbect`.
```
```suggestion
inside the array in the form of pointer to python `PyObject`.
```

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+
+Many datatype specific functions are defined within a C structure called
+:c:type:`PyArray_ArrFuncs`, which is part of each ``dtype`` instance and
+has a similarity to Pythons ``PyNumberMethods``.
```
```suggestion
has a similarity to Python's ``PyNumberMethods``.
```
Can we make this work as a link?

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+
+Although not extensively used outside of NumPy itself, the currently
+``PyArray_Descr`` is a public structure.
+This is especially also true for the ``ArrFunctions`` structure stored in
```
```suggestion
This is especially also true for the ``PyArray_ArrFuncs`` structure stored in
```

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As a side note to the above datatype implementation: unlike the datatypes,
+the NumPy scalars currently **do** provide a type hierarchy, including abstract
```
```suggestion
the NumPy scalars currently **do** provide a type hierarchy, including abstract
```

eric-wieser commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+
+
+Current Implementation of Casting
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
For my own reference: made it this far, will come back to the rest later.

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+   on the datatype which is cast, or in a dictionary when casting to a user
+   defined datatype.
+
+When casting (small) buffers will be used when necessary to ensure
```
```suggestion
When casting, (small) buffers are used when necessary to ensure
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+   defined datatype.
+
+When casting (small) buffers will be used when necessary to ensure
+contiguity, alignment or native byte order.
```
or -> and?

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+When casting (small) buffers will be used when necessary to ensure
+contiguity, alignment or native byte order.
+In this case first ``copyswapn`` is called to and ensures that the cast function
+can handle the input.
```
``copyswapn`` is typically responsible for this preprocessing to ensure these conditions are met for the subsequent casting machinery.

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+In this case first ``copyswapn`` is called to and ensures that the cast function
+can handle the input.
+Generally NumPy will thus perform casting as chain of the three functions
+``in_copyswapn -> castfunc -> out_copyswapn``.
```
This sentence is a bit redundant given the numbered points above, though I actually think this is more clear. Consider consolidating this section a bit.

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+by ufuncs.
+This is the mechanism that is always used for user defined datatypes.
+For most dtypes defined within NumPy itself, more specialized code is used to
+find define a function to do the actual cast
```
```suggestion
find a function to do the actual cast
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+find define a function to do the actual cast
+(defined by the private ``PyArray_GetDTypeTransferFunction``).
+This mechanism replaces most of the above mechanism and provides much faster
+casts for for example when the inputs are not contiguous in memory.
```
```suggestion
casts for example when the inputs are not contiguous in memory.
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+However, it cannot be extended by user defined datatypes.
+
+Related to casting, we currently have a ``PyArray_EquivTypes`` function which
+indicate that a *view* is sufficient (and thus no cast is necessary).
```
Is this another example of functionality that is not available to user-defined types?

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+
+Universal functions are implemented as instances of the ``numpy.UFunc`` class
+with an ordered
+list of datatype specific (based on the dtype typecode character, not datatype
```
ordered-list, datatype-specific

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+
+The main step in finding the correct inner-loop function is to call a
+:c:type:`PyUFunc_TypeResolutionFunc` which recieves the input dtypes
+(in the form of the actual input arrays)
```
```suggestion
:c:type:`PyUFunc_TypeResolutionFunc` which retrieves the input dtypes from 
the input arrays
```

seberg replied:
(added "provided")

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+The main step in finding the correct inner-loop function is to call a
+:c:type:`PyUFunc_TypeResolutionFunc` which recieves the input dtypes
+(in the form of the actual input arrays)
+and will find the full type signature (including output dtype) to be executed.
```
```suggestion
and will determine the full type signature (including output dtype) to be executed.
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+If a universal function, such as most functions in ``scipy.special``, is only
+defined for ``float32`` and ``float64`` it will currently automatically
+cast a ``float16`` silently to ``float32`` (similarly for any integer input).
+This ensures successful execution, but allows a change in the output dtype
```
is "allows" the right word? I'm not sure I understand the last sentence.

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+In general the order in which loops are registered is important.
+However, this is only reliable if all loops are added when the ufunc is first defined.
+Additional loops added when a new user datatypes is imported
+must not be sensitive to the order in which imports occur.
```
I'm not sure I'm understanding this correctly. What I gather is that the ufunc type resolution is sensitive to the ordering of the loops during the type lookup, and this is another feature of the current dtype system that complicates type resolution for new, user-defined types.

seberg replied:
Basically, order having an affect is not desirable. A new, improved design (which is necessary, this being one of the issues) should fix that problem.

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+the implementation-specific code described below.
+
+
+DType specific Implementation of the UFunc
```
```suggestion
DType-specific Implementation of the UFunc
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+Parametric datatypes may require passing
+additional information to the inner-loop function to decide how to interpret
+the data.
+This is the reason why currently no universal functions for strings dtypes
```
```suggestion
This is the reason why currently no universal functions for ``string`` dtypes
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+This is currently defined once per ufunc, regardless of the ufunc dtype signature.
+For example  ``0`` is used for ``sum``, or ``math.inf`` for ``min``.
+This works well for numerical datatypes, but is not always appropriate for other dtypes.
+In general it should be possible to provide a dtype specific identity to the
```
```suggestion
In general it should be possible to provide a dtype-specific identity to the
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+Datatype Discovery during Array Coercion
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When calling ``np.array(...)``, coercing general python object to a NumPy array,
```
```suggestion
When calling ``np.array(...)`` to coerce a general Python object to a NumPy array,
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+The input to ``np.array()`` are potentially nested Python sequences which hold
+the final elements as generic Python objects.
+NumPy has to unpack all the nested sequences and then inspect the elements.
+The final datatype is found by iterating all elements which will end up
```
```suggestion
The final datatype is found by iterating over all elements which will end up
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+(i.e. without the ``allow_pickle=True`` keyword argument).
+
+The additional existence of masked arrays and especially masked datatypes
+within Pandas has the interesting implications of interoperability.
```
```suggestion
within Pandas has interesting implications for interoperability.
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+in the foreseeable future.
+However, if in the interest of interoperability additions to the datatypes
+within NumPy are helpful, doing such additions could be an option even if
+they are not used by NumPy itself.
```
```suggestion
However, if such additions to the datatypes within NumPy would improve
interoperability they could be considered even if
they are not used by NumPy itself.
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+Discussion
+----------
+
+There have been many discussion about the current state and what a future
```
```suggestion
There have been many discussions about the current state and what a future
```

rossbar commented on doc/neps/nep-0040-legacy-datatype-impl.rst:
```diff
@@ -0,0 +1,616 @@
+The full list of these discussion is long and some are lost to time,
+the following provides a subset for more recent ones:
+
+* Draft on NEP by Stephan Hoyer after a developer meeting (was updated on the next developer meeting) https://hackmd.io/6YmDt_PgSVORRNRxHyPaNQ
```
```suggestion
* Draft NEP by Stephan Hoyer after a developer meeting (was updated on the next developer meeting) https://hackmd.io/6YmDt_PgSVORRNRxHyPaNQ
```

rossbar reviewed (commented):
Same commentary as in NEP 41 - mostly suggestions and nearly all of them can be safely ignored. A lot of interesting background info here although I think that it was actually beneficial having read NEP 41 in detail *before* reading this one as a lot of the context for the discussion (e.g. motivation) is actually better captured in NEP 41.

Also, just a heads-up - it seems like some of the comments from others (such as @stefanv) weren't incorporated even though they were marked as resolved. I noticed something similar on a different PR - you might want to go back and make sure that any accepted suggestions actually got applied/committed correctly.

As in NEP 41, make sure all instances of "user defined" -> "user-defined".

Thanks for putting this together, I certainly learned a lot about the dtype system :)

seberg committed:
Apply suggestions from code review

Co-Authored-By: Ross Barnowski <rossbar@berkeley.edu>
Co-Authored-By: Eric Wieser <wieser.eric@gmail.com>

seberg committed:
Smaller changes, mainly review from Ross

mattip commented:
Putting this in so it can be better reviewed and discussed.
